---
title: "Workshop Genome Wide CNVs in the UKB"
author: "Simone Montalbano"
output:
  html_document:
    toc: true
    theme: cerulea
---

**REMEMBER!** If you run the script manually do not execute chunks marked with 
`# NOT RUN`.

```{r setup, include = F}
# NOT RUN

knitr::opts_chunk$set(warning = FALSE, echo = TRUE, eval = TRUE,
                      include = TRUE, results = "markup", cache = T)
```

```{r remove cache, eval = F, include = F}
# NOT RUN 

# not eval, run manually to remove the cached files
unlink('./analysis_cache', recursive = TRUE)
unlink('./analysis_files', recursive = TRUE)
```

```{r compile html, eval = F, include = F}
# NOT RUN

# not eval, run manually to compile into an HTML
rmarkdown::render("analysis.Rmd")
```


```{r QCtreeCNV, eval = F, include = F}
# NOT RUN

# install QCtreeCNV package from GitHub if needed
devtools::install_github("sinomem/QCtreeCNV")
```

# Data and libraries

## Libs

Load necessary libraries.

```{r libs, cache = F}
library(data.table)
library(ggplot2)

# change this to the actual path on your system
setwd('~/Documents/NSHG_CNVs_workshop_2024/01_homework/')
```


# CNV table

Load simulated CNVs data.table.

```{r load CNVs}
cnvs <- fread('cnvs.tsv')
```


# Exploratory analysis

## CNV table format

The table we loaded contains all the minimal information
to define a set of CNVs, chromosome, start, end, genotype (GT)
and/or Copy Number (CN), and sample ID.

Deletions have a genotpye of 1 and a copy number of 1 or 0, while duplications
have a genotype of 2 and a copy number of 3 or more.

```{r cnvs 1}
cnvs[1:5]
```

## Visualise CNVs, export BED file for IGV

IGV is probably the best tool to visualise CNVs. The view can get complex with
high numbers, but we'll show ways to simplify that.
The most basic file format compatible with IGV is the BED4. We can
use this format to visualise a sample of our CNV table.

```{r}
# a minimal BED4 requires chr, start, end and ID. Tab as separator and
# no header

# This is a subset of 100,000 CNVs to make visualisation easier
fwrite(cnvs[sample(1:.N, 10000), .(chr, start, end, GT)],
       '../03_bed_files/cnvs_sample.bed', sep ='\t', col.names = F)
# these are all CNVs
fwrite(cnvs[, .(chr, start, end, GT)],
       '../03_bed_files/cnvs.bed', sep ='\t', col.names = F)
```

Now open IGV and load the first BED file as a track to explore how genome wide
CNVs looks like.


## Length and number of SNPs

One of the first thing we can explore are the CNVs length
and number of SNPs distributions. Length is correlated to the
number of SNPs but not perfectly, this is because the SNP density
is not constant across the array.

```{r length 1}
cnvs[, length := end - start + 1]
cnvs[, centre := round(start + length/2)]

# Length is easier to visualize in Mbp
ggplot(cnvs) + geom_density(aes(x = length/1000000)) + theme_bw()
ggplot(cnvs) + geom_histogram(aes(x = length/1000000)) + theme_bw()

ggplot(cnvs) + geom_density(aes(x = numsnp)) + theme_bw()
ggplot(cnvs) + geom_histogram(aes(x = numsnp)) + theme_bw()
```

Both distributions have a very long tail that compress most
information on the left. We can filter extreme value or use the
logarithm.

```{r length 2}
# filter out CNVs larger than 2.5Mbp
ggplot(cnvs[length <= 2500000]) +
  geom_density(aes(x = length/1000000)) + theme_bw()

# take the log
ggplot(cnvs) + geom_density(aes(x = log(length, 10))) + theme_bw()
ggplot(cnvs) + geom_histogram(aes(x = log(length, 10))) + theme_bw()
```

Log10 looks better and we don't need to exclude any CNV.


## Differences between deletions and duplications?

Now we can do the same but separate deletions and duplications
in order to see if there are differences in the distributions.

```{r legth 3}
ggplot(cnvs) +
  geom_density(aes(x = log(length, 10), colour = as.character(GT)), alpha = 0.5) +
  theme_bw()

ggplot(cnvs) +
  geom_density(aes(x = log(numsnp, 10), colour = as.character(GT)), alpha = 0.5) +
  theme_bw()
```

Before interpreting any results, remember this is simulated data. However it
is simulated from real CNVs, and these results follows what we usually see.
Duplication tends to be larger than deletions and the distributions are not as
smooth as one might expect.


## CNVs density across the genome

We can have an idea of how CNVs are distributed across the chromosomes
using the centre position.

```{r}
ggplot(cnvs[chr %in% c(1, 2, 6, 9, 10, 12, 14, 21, 22), ]) +
  geom_density(aes(x = centre/1000000, fill = as.character(GT)), alpha = 0.3) +
  facet_wrap(~chr, scales = 'free')
```

I selected a handful of chromosomes that have similar outputs to "the real thing"
in this specific simulation.



# Homework

Now that you have an idea on how CNVs looks like and how they are distributed in
the genome you can think how you would analyse them.

We have prepared an excersise for that. In this table you will find the samples list
and one columns with the phenotype state, 1 for cases and 0 for controls.

There are four loci in the genome associated with the phenotype, can you find them?

Some hints / suggestions:

- Treating each CNV as an exposure is infeasible, what about grouping them?
- Maybe set some windows across the genome?
- What about genes?

```{r}
pheno <- fread('pheno.txt')
pheno[1:5]
```

If your strategy works you will find a signal somewhere in chromosomes
5, 8, 13 and 21.

Good luck! ;)
