---
title: "Workshop Genome Wide CNVs in the UKB"
author: "Simone Montalbano"
output:
  html_document:
    toc: true
    theme: cerulea
---

```{r setup, include = F}
knitr::opts_chunk$set(warning = FALSE, echo = TRUE, eval = TRUE,
                      include = TRUE, results = "markup", cache = T)
```

```{r remove cache, eval = F, include = F}
# not eval, run manually to remove the cached files
unlink('./analysis_cache', recursive = TRUE)
unlink('./analysis_files', recursive = TRUE)
```


# Data and libraries

Load necessary libraries and CNValidatron package.

```{r libs, cache = F}
library(data.table)
library(ggplot2)
setwd('../CNValidatron_fl')
devtools::load_all()
setwd('../NSHG_CNVs_workshop_2024')
```

Load CNVs data.table.

```{r load CNVs}
cnvs <- fread('cnvs.tsv')
```


# Exploratory analysis

## CNV table format

The table we loaded contains all the minimal information
to define a set of CNVs, chromosome, start, end, genotype (GT)
and/or Copy Number (CN), and sample ID.

```{r cnvs 1}
cnvs[1:5]
```

## Visualise CNVs, export BED file for IGV

IGV is probably the best tool to visualise CNVs. It can get complex with
high number, but we'll show ways to simplify that.
The most basic file format compatible with IGV is the BED4. We can
use this format to visualise a sample of our CNV table.

```{r}
# a minimal BED4 requires chr, start, end and ID. Tab as separator and
# no header
fwrite(cnvs[sample(1:.N, 10000), .(chr, start, end, GT)],
       'cnvs_sample.bed', sep ='\t', col.names = F)
```


## Length and number of SNPs

One of the first thing we can explore are the CNVs length
and number of SNPs distributions. Length is correlated to the
number of SNPs but not perfectly, this is because the SNP density
is not constant across the array.

```{r length 1}
cnvs[, length := end - start + 1]
cnvs[, centre := round(start + length/2)]

# Length is easier to visualize in Mbp
ggplot(cnvs) + geom_density(aes(x = length/1000000)) + theme_bw()

ggplot(cnvs) + geom_density(aes(x = numsnp)) + theme_bw()
```

Both distributions have a very long tail that compress most
information on the left. We can filter extreme value or use the
logarithm.

```{r length 2}
# filter out CNVs larger than 2.5Mbp
ggplot(cnvs[length <= 2500000]) +
  geom_density(aes(x = length/1000000)) + theme_bw()

# take the log
ggplot(cnvs) + geom_density(aes(x = log(length, 10))) + theme_bw()
```

Log10 is better and we don't need to exclude any CNV.


## Differences between deletions and duplications?

Now we can do the same but separate deletions and duplications
in order to see if there are differences in the distributions.

```{r legth 3}
ggplot(cnvs) +
  geom_density(aes(x = log(length, 10), colour = as.character(GT)), alpha = 0.5) +
  theme_bw()

ggplot(cnvs) +
  geom_density(aes(x = log(numsnp, 10), colour = as.character(GT)), alpha = 0.5) +
  theme_bw()
```

Before interpreting any results, remember this is simulated data. However it
is simulated from real CNVs, and these results follows what we usually see.
Duplication tends to be larger than deletions and the distributions are not as
smooth as one might expect.


## CNVs density across the genome

We can have an idea of how CNVs are distributed across the chromosomes
using the centre position.

```{r}
ggplot(cnvs[chr %in% c(1, 2, 6, 9, 10, 12, 14, 21, 22), ]) +
  geom_density(aes(x = centre/1000000, fill = as.character(GT)), alpha = 0.3) +
  facet_wrap(~chr, scales = 'free')
```

I selected a handful of chromosomes that have similar outputs to "the real thing"
in this specific simulation.


## Other?



# CNVRs

```{r}

```


# Access Ensembl using biomaRt

```{r}

```


# Genomic enrichment analysis

```{r}

```

